drafts/str4d-overwinter-activation/draft1.rst

```diff
@@ -0,0 +1,237 @@
+version has been deprecated, they can reasonably assume that all node operators on the network either support
+the upgraded rules, or have explicitly chosen not to follow them.
+
+However, this behaviour is not sufficient for performing network upgrades. A globally-understood on-chain
+activation mechanism is necessary so that nodes can unambiguously know at what point the changes from an
+upgrade come into effect (and can enforce consensus rule changes, for example).
+
+
+Specification
+=============
+
+The following constants are defined for every network upgrade:
+
+BRANCH_ID
+  A globally-unique non-zero 32-bit identifier.
+
+  Implementations MAY use a value of zero in branch ID fields to indicate the absence of any upgrade (i.e.
+  that the Sprout consensus rules apply).
+
+ACTIVATION_HEIGHT
+  The non-zero block height at which the network upgrade rules will come into effect, and be enforced as part
+  of the blockchain consensus.
+
+  For removal of ambiguity, the block at height ``ACTIVATION_HEIGHT - 1`` is subject to the pre-upgrade
+  consensus rules, and would be the last common block in the event of a persistent pre-upgrade branch.
+
+  It MUST be greater than the value of ``DEPRECATION_HEIGHT`` in the last software version that will not
+  contain support for the network upgrade. It SHOULD be chosen to be reached approximately three months after
+  the first software version containing support for the network upgrade is released, for the following reason:
+
+  - As of the time of writing (the 1.0.15 release), the release cycle is six weeks long, and nodes undergo
+    auto-senescence 16 weeks after release. Thus, if version ``X`` contains support for a network upgrade,
+    version ``X-1`` will deprecate 10 weeks after the release of version ``X``, which is about 2.3 months. A
+    three-month window provides ample time for users to upgrade their nodes after auto-senescence, and
+    re-integrate into the network prior to activation of the network upgrade.
+
+The relationship between ``BRANCH_ID`` and ``ACTIVATION_HEIGHT`` is many-to-one: it is possible for many
+distinct branches to descend from the same parent block (and thus have the same ``ACTIVATION_HEIGHT``), but a
+specific branch can only have one parent block. Concretely, this means that if the ``ACTIVATION_HEIGHT`` of a
+network upgrade is changed for any reason (e.g. security vulnerabilities or consensus bugs are discovered),
+the ``BRANCH_ID`` MUST also be changed.
+
+Activation mechanism
+--------------------
+
+A blockchain is defined as invalid if, within the set of all network upgrades that have activated in the past
+(or will activate in future) on that chain, an ``ACTIVATION_HEIGHT`` is repeated. Note that this does not
+require ``ACTIVATION_HEIGHT`` to be globally unique, or even locally unique; multiple network upgrades can
+occur in parallel, as long as they are non-overlapping (only one will activate on any given chain).
```

Comment:
If two branches on the same chain have the same activation height, which one takes precedence? We can't enforce an on-chain ordering without having at least one block of separation, and from a practical perspective you should just release the two sets of rules under one branch ID.

The sole exception to this may be in a code fork with a brand-new chain with new genesis block, if all pre-existing upgrades are set to `ALWAYS_ACTIVE`. But that is an implementation detail to avoid re-coding all the consensus rules, and essentially amounts to saying that the last upgrade in the array is the one that applies. I highly doubt that we want to enable this kind of implementation-specific detail to creep into the live consensus rules.

Reply:
I agree, the whole part from "Note that" is not clear. In particular, I'm not sure how you define chain,  I guess just any chain of blocks?  But then what do you mean by two branches on the same chain in your comment? That sounds like you're also calling a tree a chain.

Reply:
@str4d I agree that we don't want to intentionally have branches with the same activation height. The point was that it's headache-inducing to figure out (at least from this paragraph) whether the definitions allow it. They either explicitly should or expicitly shouldn't. But since we can't actually *prevent* multiple branches with the same activation height, they explicitly should &mdash; and indeed the last paragraph of the previous section does so, but this paragraph just muddies the waters.

TL;DR: delete the paragraph.